keanu-project/src/main/java/io/improbable/keanu/network/ModelComposition.java

```diff
@@ -0,0 +1,98 @@
+package io.improbable.keanu.network;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import io.improbable.keanu.vertices.ProxyVertex;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.VertexLabel;
+
+public final class ModelComposition {
+
+    private ModelComposition() { }
+
+    /**
+     * Compose one model within a wider model.
+     * <p>
+     *     This function will take a previously constructed BayesNet and do a number of things:
+     *     - Hook up any Proxy Vertices to the parents specified in the inputVertices
+     *     - Increment the depth for all "internal" nodes and the BayesNet itself
+     *     - Pass back any specified output nodes from the Bayesnet (keeping them at the outer depth)
+     *
+     *     Output nodes will have be returned unlabelled
+     * </p>
+     * @param bayesianNetwork The Bayesian Network to compose in to the current model
+     * @param inputVertices The mapping from Proxy label to actual input vertex
+     * @param desiredOutputs The set of labels we wish to output from the supplied BayesNet
+     * @return A map of Labels to Output Vertices
+     */
+    public static Map<VertexLabel, Vertex> createModelVertices(BayesianNetwork bayesianNetwork,
+                                                               Map<VertexLabel, Vertex> inputVertices,
+                                                               List<VertexLabel> desiredOutputs) {
+        Map<VertexLabel, Vertex> outputMap = extractOutputs(bayesianNetwork, desiredOutputs);
+        increaseDepth(bayesianNetwork, outputMap);
+        checkAndLinkInputs(bayesianNetwork, inputVertices);
+        cleanOutputLabels(outputMap);
+
+        return outputMap;
+    }
+
+    private static Map<VertexLabel, Vertex> extractOutputs(BayesianNetwork bayesianNetwork,
+                                                           List<VertexLabel> desiredOutputs) {
+        if (desiredOutputs.size() == 0) {
+            throw new IllegalArgumentException("At least one output must be specified");
+        }
+
+        Map<VertexLabel, Vertex> outputMap = new HashMap<>();
+
+        for (VertexLabel label: desiredOutputs) {
+            Vertex v = bayesianNetwork.getVertexByLabel(label);
+            if (v == null) {
+                throw new IllegalArgumentException("Unable to find Output Vertex: " + label);
+            }
+            outputMap.put(label, v);
+        }
+
+        return outputMap;
+    }
+
+    private static void increaseDepth(BayesianNetwork bayesianNetwork, Map<VertexLabel, Vertex> outputVertices) {
+        VertexId newPrefix = new VertexId();
+        bayesianNetwork.incrementDepth();
+        bayesianNetwork.getAllVertices().stream()
```

Comment:
Could we pass a filter (and maybe also `increaseDepth`) to `BayesianNetwork` instead of asking for all the vertices?

Reply:
Probably easiest to chat on this - it would be possible to do this, but I'm not sure it would necessarily be clearer (and we're just returning a list reference, so there's no perf concern here).

I wanted to keep all this logic together for now as really this process is more of something we'll do to a "model" rather than an "instantiation" (ie a BayesNet) so in future this code won't likely interact with a BayesNet at all anyway...